### ISSUE TITLE
Undefined handlers cause error when unsafe-eval is restricted by the CSP header

### ISSUE THREAD

**Issue by KarsonAlford** [aspnet/jquery-ajax-unobtrusive#49]
Undefined handlers (OnBegin, OnSuccess, etc.) cause a browser evaluation error when unsafe-eval is restricted by the CSP header.

If you modify all of the text in the handlers to functions instead of JavaScript code then those handlers that contain functions will no longer be blocked by the browser (window.stop [function] vs window.stop() [code needing evaluation]).  However, the issue is that the handlers that are left blank will be blocked by the browser and cause an error.

Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive:

Related to [Issue #23](https://github.com/aspnet/jquery-ajax-unobtrusive/issues/23).

**Comment by KarsonAlford** [comment]
**Explanation:**
The getFunction method first tries to find the function in the DOM and falls back to eval via Function.constructor.  Currently if one of the handlers (OnBegin, OnSuccess, etc.) is not defined then null is passed into getFunction which is not a function and therefore is evaluated and a function created that returns null. 


**Possible Fix:**

    function getFunction(code, argNames) {
        if (!code) return function() { return null; };



**Current Options:**
Option 1) Add a NOOP function for all the handlers that you are not using.

`window.noop = function() { return null; };`

OnBegin="noop"
OnSuccess="noop"


Option 2) Replace Function.apply so that the eval is not possible.  Inspired by [this blog post](https://statuscode.ch/2017/03/CSP-unsafe-eval-and-jquery).

    Function.apply = function (element, args) {
        if (args && args.length && args.length !== 0) {
            var code = args[args.length - 1];
            if (code) {
                console.log('JavaScript evaluation prevented - ' + code);
            }
        }
        return function () { return null; };
    };

**Comment by AdamWillden** [comment]
@KarsonAlford I completely agree and cannot understand why eval is used even in cases where there's nothing to evaluate.

Please @mkArtakMSFT, @javiercn reconsider your stance on this. There's no need to depreciate existing features [as was described previously](https://github.com/aspnet/jquery-ajax-unobtrusive/issues/23#issuecomment-365404265) in order to support the avoidance of eval.

**Comment by jumpingjackson** [comment]
I tried applying the changes in https://github.com/aspnet/jquery-ajax-unobtrusive/pull/48 to my copy, but i still get CSP violation.

**Comment by KarsonAlford** [comment]
Can you post your example?

This fix is to stop empty handlers causing a CSP violation.  If you are getting past !code then I assume you have something in a handler that is not a function in the DOM and therefore is being evaluated.

**Comment by jumpingjackson** [comment]
Sorry l thought it was a fix for what I now know is the other bug that they
said they won’t fix.  I’m trying to call a function in a .js file on
success and getting the unsafe eval csp error.

**Comment by calaluis** [comment]
> **Explanation:**
> The getFunction method first tries to find the function in the DOM and falls back to eval via Function.constructor. Currently if one of the handlers (OnBegin, OnSuccess, etc.) is not defined then null is passed into getFunction which is evaluated and a function created that returns null.
> 
> **Possible Fix:**
> 
> ```
> function getFunction(code, argNames) {
>     if (!code) return function() { return null; };
> ```
> 
> **Current Options:**
> Option 1) Add a NOOP function for all the handlers that you are not using.
> 
> `window.noop = function() { return null; };`
> 
> OnBegin="noop"
> OnSuccess="noop"
> 
> Option 2) Replace Function.apply so that the eval is not possible. Inspired by [this blog post](https://statuscode.ch/2017/03/CSP-unsafe-eval-and-jquery).
> 
> ```
> Function.apply = function (element, args) {
>     if (args && args.length && args.length !== 0) {
>         var code = args[args.length - 1];
>         if (code) {
>             console.log('JavaScript evaluation prevented - ' + code);
>         }
>     }
>     return function () { return null; };
> };
> ```

Hello dear,

First of all, thanks to find a possible solution to this big problem, for an extremely used library in ASP.NET MVC.

I can inform you that option 2 has been implemented, within which it throws the following CSP error towards the Jquery library:

![image](https://user-images.githubusercontent.com/33674615/99852119-150d1b00-2b5f-11eb-8b14-76b1d4ed5bc9.png)


Stay tuned to your comments.

**Comment by KarsonAlford** [comment]
See the JS that I am using below.  It also prevents $.globalEval which I didn't document in my original post.

I hope this helps.

disableEval.js
```
(function ($) {
    "use strict";

    // replace apply for Function.constructor
    Function.apply = function (element, args) {
        if (args && args.length && args.length !== 0) {
            var code = args[args.length - 1];
            if (code) {
                console.log('JavaScript evaluation prevented - ' + code);
            }
        }

        return function () { return null; };
    };

    $(function () {
        $.globalEval = function (code) { console.log('jQuery evaluation prevented - ' + code); };
    });
}(jQuery));
```

**Comment by calaluis** [comment]
Thank you very much @KarsonAlford , it worked perfect.

Now to investigate other libraries that also give problems with the CSP, such as the BlockUI.

**Comment by KAJOOSH** [comment]
After a few years, don't you want to think fundamentally to solve this problem?
